File: Data Structures & Algorithms/add-two-numbers/submission-3.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        current = dummy
        carry_in = 0

        while l1 or l2 or carry_in:
            l1_val = l1.val if l1 else 0
            l2_val = l2.val if l2 else 0

            summ = l1_val + l2_val + carry_in

            carry_in = summ // 10
            summ = summ % 10

            current.next = ListNode(summ)
            current = current.next

            if l1:
                l1 = l1.next
            if l2:
                l2 = l2.next

        return dummy.next

File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-3.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode* dummy = new ListNode();
        ListNode* current = dummy;

        while (list1 != nullptr && list2 != nullptr) {
            int val1 = list1->val;
            int val2 = list2->val;

            if (val1 <= val2) {
                current->next = list1;
                list1 = list1->next;
            } else {
                current->next = list2;
                list2 = list2->next;
            }

            current = current->next;
        }

        while (list1 != nullptr) {
            current->next = list1;
            list1 = list1->next;
            current = current->next;
        }

        while (list2 != nullptr) {
            current->next = list2;
            list2 = list2->next;
            current = current->next;
        }

        return dummy->next;
    }
};
